test/integration/agent/gobgp_test.go

```diff
@@ -0,0 +1,436 @@
+		RouterID:   routerID2,
+		ListenPort: listenPort2,
+	}
+	server3GlobalConfig := &bgp.GlobalConfig{
+		ASN:        uint32(asn3),
+		RouterID:   routerID3,
+		ListenPort: listenPort3,
+	}
+
+	server1 := gobgp.NewGoBGPServer(server1GlobalConfig)
+	server2 := gobgp.NewGoBGPServer(server2GlobalConfig)
+	server3 := gobgp.NewGoBGPServer(server3GlobalConfig)
+
+	ctx := context.TODO()
+	t.Log("Starting all BGP servers")
+	require.NoError(t, server1.Start(ctx))
+	require.NoError(t, server2.Start(ctx))
+	require.NoError(t, server3.Start(ctx))
+	t.Log("Started all BGP servers")
+
+	ipv4Server1Config := bgp.PeerConfig{
+		BGPPeer: &v1alpha1.BGPPeer{
+			Address:                    "127.0.0.1",
+			Port:                       ptr.To(int32(1179)),
+			ASN:                        61179,
+			MultihopTTL:                ptr.To(int32(2)),
+			GracefulRestartTimeSeconds: ptr.To(int32(120)),
+		},
+	}
+	ipv6Server1Config := bgp.PeerConfig{
+		BGPPeer: &v1alpha1.BGPPeer{
+			Address:                    "::1",
+			Port:                       ptr.To(int32(1179)),
+			ASN:                        61179,
+			MultihopTTL:                ptr.To(int32(2)),
+			GracefulRestartTimeSeconds: ptr.To(int32(120)),
+		},
+	}
+	ipv4Server2Config := bgp.PeerConfig{
+		BGPPeer: &v1alpha1.BGPPeer{
+			Address:                    "127.0.0.1",
+			Port:                       ptr.To(int32(2179)),
+			ASN:                        62179,
+			MultihopTTL:                ptr.To(int32(3)),
+			GracefulRestartTimeSeconds: ptr.To(int32(130)),
+		},
+	}
+	ipv6Server3Config := bgp.PeerConfig{
+		BGPPeer: &v1alpha1.BGPPeer{
+			Address:                    "::1",
+			Port:                       ptr.To(int32(3179)),
+			ASN:                        63179,
+			MultihopTTL:                ptr.To(int32(1)),
+			GracefulRestartTimeSeconds: ptr.To(int32(140)),
+		},
+	}
+
+	t.Log("Adding BGP peers for BGP server1")
+	require.NoError(t, server1.AddPeer(ctx, ipv4Server2Config))
+	require.NoError(t, server1.AddPeer(ctx, ipv6Server3Config))
+	t.Log("Added BGP peers for BGP server1")
+
+	t.Log("Adding BGP peers for BGP server2")
+	require.NoError(t, server2.AddPeer(ctx, ipv4Server1Config))
+	t.Log("Added BGP peers for BGP server2")
+
+	t.Log("Adding BGP peers for BGP server3")
+	require.NoError(t, server3.AddPeer(ctx, ipv6Server1Config))
+	t.Log("Added BGP peers for BGP server3")
+
+	t.Log("Getting peers of BGP server1 and verifying them")
+	assert.Eventually(t, func() bool {
+		peers, err := server1.GetPeers(ctx)
+		if err != nil {
+			return false
+		}
+		expectedPeers := sets.New[string]("::1-63179", "127.0.0.1-62179")
+		gotPeers := sets.New[string]()
+		for _, peer := range peers {
+			if peer.SessionState != bgp.SessionEstablished {
+				return false
+			}
+			t.Log(peer)
+			gotPeers.Insert(fmt.Sprintf("%s-%d", peer.Address, peer.ASN))
+		}
+		if !expectedPeers.Equal(gotPeers) {
+			return false
+		}
+		return true
+	}, 30*time.Second, time.Second)
+	t.Log("Got peers of BGP server1 and verified them")
```

Comment:
would it make sense to define a helper function here to retrieve peers and avoid the verbosity (given that we repeat the same logic below for the other servers):
```go
getPeers := func(server bgp.Interface) sets.Set[string] {
        // ...
}
assert.EventuallyWithT(t, func(t *assert.CollectT) {
        assert.Equal(t, sets.New[string]("::1-63179", "127.0.0.1-62179"), getPeers(server1))
}
```

Note: I think this is better than defining a `checkPeers(t *testing.T, expectedPeers sets.Set[string], server bgp.Interface)` function because in case of a test failure the error message (in particular the line numbers which are reported) is more helpful

Reply:
Make sense! Then we can know the exact place where the error message is generated.